pkg/pmem-csi-driver/registryserver_test.go

```diff
@@ -0,0 +1,196 @@
+		registryClient = registry.NewRegistryClient(registryClientConn)
+	})
+
+	AfterEach(func() {
+		if registryServer != nil {
+			nbServer.ForceStop()
+			nbServer.Wait()
+		}
+		os.Remove(registryServerEndpoint)
+		if registryClientConn != nil {
+			registryClientConn.Close()
+		}
+	})
+
+	Context("Registry API", func() {
+		var (
+			nodeId      = "pmem-test"
+			registerReq = registry.RegisterControllerRequest{
+				NodeId:   nodeId,
+				Endpoint: controllerServerEndpoint,
+				Capacity: map[string]uint64{
+					"fsdax": 1 * 1024 * 1024, // 1GB
+				},
+			}
+
+			unregisterReq = registry.UnregisterControllerRequest{
+				NodeId: nodeId,
+			}
+		)
+
+		It("Register node controller", func() {
+			Expect(registryClient).ShouldNot(BeNil())
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			_, err := registryClient.RegisterController(ctx, &registerReq)
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = registryServer.GetNodeController(nodeId)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Unregister node controller", func() {
+			Expect(registryClient).ShouldNot(BeNil())
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			_, err := registryClient.RegisterController(ctx, &registerReq)
+			Expect(err).NotTo(HaveOccurred())
+
+			ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			_, err = registryClient.UnregisterController(ctx, &unregisterReq)
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = registryServer.GetNodeController(nodeId)
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("Unregister non existing node controller", func() {
+			Expect(registryClient).ShouldNot(BeNil())
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			_, err := registryClient.UnregisterController(ctx, &unregisterReq)
+			Expect(err.Error()).To(ContainSubstring("No entry with id '" + nodeId + "' found in registry"))
+		})
+	})
+
+	Context("Registry Security", func() {
+		var (
+			evilEndpoint = "unix:///tmp/pmem-evil.sock"
+			ca           = os.ExpandEnv("${TEST_WORK}/pmem-ca/ca.pem")
+			evilCA       = os.ExpandEnv("${TEST_WORK}/evil-ca/ca.pem")
+			evilCert     = os.ExpandEnv("${TEST_WORK}/evil-ca/pmem-node-controller.pem")
+			evilKey      = os.ExpandEnv("${TEST_WORK}/evil-ca/pmem-node-controller-key.pem")
+		)
+
+		// This covers different scenarios for connections to the OIM registry.
+		cases := []struct {
+			name, ca, cert, key, errorText string
+		}{
+			//{"should be accepted", ca, caCert, caKey, "host-0: no address registered"},
+			{"registry should detect man-in-the-middle", ca, evilCert, evilKey, "authentication handshake failed: remote error: tls: bad certificate"},
+			{"client should detect man-in-the-middle", evilCA, evilCert, evilKey, "transport: authentication handshake failed: x509: certificate signed by unknown authority"},
+		}
```

Comment:
Can you add also the two cases where the certificate is valid, but the common name is wrong?

Reply:
As suggested, I added a case where "the certificate is valid, but the common name is wrong".

Reply:
I mentioned *two* cases ;-} The case "server should detect wrong peer" is still missing.

Reply:
wrong peers you mean, registry server shall not allow connections from non "pmem-node-controllers"?

Reply:
I don't know - you tell me which connections it should accept. Those, and nothing, else, should be accepted.

But yeah, only allowing connections from "pmem-node-controllers" sounds right.

Reply:
Done